### ISSUE TITLE
Completed torrents not being deleted

### ISSUE THREAD

**Issue by jamaroney** [barbequesauce/Watcher3#192]
This has always been an intermittent problem since I started using Watcher about 2 years ago, but now it's constant.  

"Remove Finished Torrents" box is checked
"Stop seeding when ratio X is reached" - set to any positive whole number

When the ratio is reached, the torrent stops seeding, but isn't removed. A while back it actually worked on occasion (but by no means the majority of times), but in the last few months it has stopped working completely.

I've always used the latest version of Watcher (auto update). I'm currently on FreeBSD 12.1 w/ Transmission 3.0, but the problem has existed w/ earlier versions of both. 

I don't know if it's relevant, but I have the same feature enabled in Sonarr w/ absolutely no problems.

**Comment by barbequesauce** [comment]
I've been chasing this since I upgraded to Tx 3.0. Did you see it on earlier versions?

**Comment by jamaroney** [comment]
> I've been chasing this since I upgraded to Tx 3.0. Did you see it on earlier versions?

Yes. As I noted in my earlier post, this also occurred in Transmission 2.9.* versions.

**Comment by barbequesauce** [comment]
Can you grab a debug log please? I'd like to compare to mine...

**Comment by jamaroney** [comment]
The next time I download a movie I'll send the debug log

**Comment by jamaroney** [comment]
Here's a debug log. I was downloading "Honest Thief" yesterday. I set the ratio for the torrent to 1, and Transmission completed its seeding (status - "Finished") sometime early this morning while I was sleeping. I included only the later part of the log (while it was still seeding), until after its completion.
[debug.txt](https://github.com/barbequesauce/Watcher3/files/5862069/debug.txt)

**Comment by barbequesauce** [comment]
Just added some logging around the decision process - if you can update and then post another log once you've snatched/finished a movie please...

**Comment by jamaroney** [comment]
OK

**Comment by barbequesauce** [comment]
Noticed this in the logs - does that make sense to you? Not related to your concern, just interesting...

```
NFO [2021-01-26 06:52:11,531] core.plugins.execute.263: encode.py - b"/usr/local/bin/python3: can't open file '/mnt/HDD/extensions/watcher/plugins/finished/encode.py': [Errno 2] No such file or directory"
WARNING [2021-01-26 06:52:11,532] core.plugins.execute.267: encode.py - Execution failed. Exit code 2.
```

The develop branch has some additional logging, if you wouldn't mind switching and trying there?

**Comment by jamaroney** [comment]
I have no idea what those log messages mean. I didn't have any plugin scripts set to run.

How do I convert to develop branch without losing all my current settings? I originally installed Watcher via a ZIP file, not through git.

**Comment by barbequesauce** [comment]
Not sure... anytime I've seen someone make that jump they do a new install from GitHub and copy over the ini/db..

**Comment by jamaroney** [comment]
I made the switch to develop branch (far easier than I feared), and then downloaded "Space Sweepers" (2021). Same results - it reached its max seeding ratio (marked as "Finished" in transmission), and is still there.  Attached is the debug log.
[log.txt](https://github.com/barbequesauce/Watcher3/files/6024996/log.txt)

**Comment by barbequesauce** [comment]
OK - update to that please.

The original added log lines pointed at the issue I expected, but I'm guessing that we are handling the config option badly - either loading from the config or referencing here. This last bit of logging should clear that up.

**Comment by jamaroney** [comment]
Updated to newest develop branch. Downloaded same movie as before (got rid of all traces of the old  movie files and torrent beforehand). Attached is the new log.

On a similar note: How would one confirm if a develop branch has been installed
[log.txt](https://github.com/barbequesauce/Watcher3/files/6029034/log.txt)
? I did a clean install, but the "Current version hash" on the "Server" page shows the latest master branch. Is that normal?

**Comment by barbequesauce** [comment]
So that solves that - the torrent client shows the torrent in status "seeding", not "finished":

`INFO [2021-02-23 06:56:18,387] core.scheduler.check_torrents.384: Torrent status 28bf3e6c36005869ed1f89e1ffd640be076856e4 [seeding] (b'Space Sweepers (2021) [720p] [WEBRip] [YTS.MX]'); removetorrents: True`

To confirm branch: git status

```
watcher@pi:/opt/watcher3 $ git status
On branch develop
Your branch is up to date with 'origin/develop'.

nothing to commit, working tree clean
watcher@pi:/opt/watcher3 $
```

**Comment by jamaroney** [comment]
Yet the torrent in Transmission has reached the correct seed ratio and states that the torrent is "Finished," and has definitely stopped seeding. So, is there something in "transmission.py" that needs to be changed?

Since I don't use git, is there any other way to confirm develop branch? It's not important, just curious....

**Comment by barbequesauce** [comment]
looks like the transmission rpc spec does not update the status to `finished` but instead uses a flag to indicate `isFinished` ([see](https://github.com/transmission/transmission/blob/9c5e6f7ea61ad4e7708756b7a9b7c299936117f4/extras/rpc-spec.txt#L206). I have some homework to do...

**Comment by jamaroney** [comment]
Thanks for your diligence with this.

**Comment by barbequesauce** [comment]
That seems to work here...

**Comment by jamaroney** [comment]
Installed the patched version. Downloaded two torrents, and both are "Finished." Neither have been removed.

**Comment by jamaroney** [comment]
Out of the blue, Watcher actually deleted a finished torrent!  But afterward, it went back to its usual mode of not removing. Here's the essential log files for both:

INFO [2021-03-30 01:49:28,267] core.scheduler.check_torrents.384: Torrent status b741363904a505a5129b03c466d15aece90e7fdc [finished] (b'The.Insider.1999.1080p.BluRay.x265-RARBG'); removetorrents: True
INFO [2021-03-30 01:49:28,269] core.scheduler.check_torrents.386: Check if we know finished torrent b741363904a505a5129b03c466d15aece90e7fdc and is postprocessed (b'The.Insider.1999.1080p.BluRay.x265-RARBG')
INFO [2021-03-30 01:49:28,302] core.scheduler.check_torrents.388: Yes, now we remove the torrent
INFO [2021-03-30 01:49:28,304] core.downloaders.Transmission.cancel_download.188: Cancelling download # b741363904a505a5129b03c466d15aece90e7fdc in Transmission.
__________________________________________________________________________
INFO [2021-03-30 08:49:28,306] core.scheduler.check_torrents.384: Torrent status a37f3eeab6612bcb6451e5e9061a4ba796b3f641 [finished] (b'www.lyrics.lol   -   The.Father.2020.1080p.WEBRip.x264-MOTHER'); removetorrents: True
INFO [2021-03-30 08:49:28,308] core.scheduler.check_torrents.386: Check if we know finished torrent a37f3eeab6612bcb6451e5e9061a4ba796b3f641 and is postprocessed (b'www.lyrics.lol   -   The.Father.2020.1080p.WEBRip.x264-MOTHER')
INFO [2021-03-30 08:49:28,324] core.scheduler.check_torrents.391: No, did not find the torrent as finished
